pkg/manifest/manifest.go

```diff
@@ -5,36 +5,95 @@
 package manifest
 
 import (
-	"encoding"
+	"context"
 	"errors"
-	"net/http"
 
+	"github.com/ethersphere/bee/pkg/storage"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
-// ErrNotFound is returned when an Entry is not found in the manifest.
-var ErrNotFound = errors.New("manifest: not found")
+const DefaultManifestType = ManifestSimpleContentType
+
+var (
+	// ErrNotFound is returned when an Entry is not found in the manifest.
+	ErrNotFound = errors.New("manifest: not found")
+
+	// ErrInvalidManifestType is returned when an unknown manifest type
+	// is provided to the function.
+	ErrInvalidManifestType = errors.New("manifest: invalid type")
+)
 
 // Interface for operations with manifest.
 type Interface interface {
+	// Type returns manifest implementation type information
+	Type() string
 	// Add a manifest entry to the specified path.
-	Add(string, Entry)
+	Add(string, Entry) error
 	// Remove a manifest entry on the specified path.
-	Remove(string)
-	// Entry returns a manifest entry if one is found in the specified path.
-	Entry(string) (Entry, error)
-	// Length returns an implementation-specific count of elements in the manifest.
-	Length() int
-	encoding.BinaryMarshaler
-	encoding.BinaryUnmarshaler
+	Remove(string) error
+	// Lookup returns a manifest entry if one is found in the specified path.
+	Lookup(string) (Entry, error)
+	// Store stores the manifest, returning the resulting address.
+	Store(storage.ModePut) (swarm.Address, error)
 }
 
 // Entry represents a single manifest entry.
 type Entry interface {
-	// Reference returns the address of the file in the entry.
+	// Reference returns the address of the file.
```

Comment:
so now a manifest entry only knows to represent an address and that's it?

Reply:
Yes, that is the change (that I mentioned in the description of this PR).

Name, as you previously noted should be removed as it is a bit pointless in this structure, can just be extracted from the path using `filepath.Base` at any point. As for the headers go, only header that is being added is the `Content-Type` header during upload (again, based on file name using `mime.TypeByExtension`).

Both of these things were also being saved in the actual file on the Swarm in the special `entry.Entry` object, with reference to `entry.Metadata` which contains filename and mime type.

At the moment there is no way (at least I count not find one) to modify the name of already uploaded file after the fact. Also there is no way to add additional headers to the file, or modify existing manifest (through some API) to add additional headers.

Regarding uploading TAR collection using `/dirs` API: Marcelo and myself have been working on separate parts for "website support", and while I added initial manifest implementation (that had 'name' and 'headers') and serving content via `/bzz` API, actual expectation I had was that in that case content would be saved as is using 'bytes' API, while Marcelo used 'files' API (saving files with metadata) on the `/dirs` API side.
During the merge time we had a meeting with Rinke and Viktor regarding this, and decision was to use 'files' for storing files, meaning manifest will contain references to files, which themselves already have metadata which was initially added in manifest. Since it is already located there, I see no point in having it here also.

Now, we could change this back and say that we want to save only bytes for manifest uploaded files, in which case it is easy to return this to previous state.

P.S.
Important note for this is that the initial `mantaray` implementation does not have support for complex content inside. This can easily be changed though.
Also I would like to note that this fact was not the reason for removing fields.

Reply:
Related: https://github.com/ethersphere/bee/issues/463#issuecomment-677590074